### PR TITLE
Rana & Alix Scrabble

### PR DESCRIPTION
**Pull request by alixhami**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   We did a good job of splitting up the responsibilities of each wave.	|
|   Why do you think we chose to make the `#score` method in the `Scoring` class a **class method** instead of an **instance method**? |  The `score` method makes sense as a class method, because it is consistently used for all Players, so we don't need to create a new score instance when we use it. The Scoring class is just a calculator that doesn't need to be differentiated.  |
|   Describe an example of a test you wrote for a _nominal case_. |   We tested to make sure that `#score` will calculate the score for the word "apple" as our expected score of 9, based on the `CONSTANT_SCORE_CHART`. This is a nominal case, because it's a normal input that we would expect to get from the user.	|
|   Describe an example of a test you wrote for an _edge case_. |  We tested to make sure that `#score` would raise an `ArgumentError` if given a number instead of a string. This is an edge case because it's an unusual input that could occur.	|
|   What was your final test coverage percentage?	|   100% excluding the game class, and 88% including the game class.	|
|   Is there a specific piece of code you'd like feedback on?	|   Not particularly.	|

### CODE REVIEW

File: lib/player.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    attr_reader :name, :plays, :tiles
+
+    def initialize(name)
+      @name = name
+      @plays = []
+      @tiles = []
+    end
+
+    def play(word)
+      raise ArgumentError.new("Must enter a valid string.") if word.nil? || word.empty?
```
Comment (CheezItMan) [inline lib/player.rb:12]: What if `word` is not a String?  What if it's an Array?

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    end
+  end
+
+  describe "#highest_score_from" do
+    it "Returns the word with the highest score" do
+      words = ["chicken", "apple"]
+      Scrabble::Scoring.highest_score_from(words).must_equal "chicken"
+    end
+
+    it "Returns word with fewer letters in a tie" do
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:31]: You should also probably be testing to make sure order doesn't affect the word selected if the words are of different length.

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      words = ["land", "dig", "so"]
+      Scrabble::Scoring.highest_score_from(words).must_equal "dig"
+    end
+
+    it "Returns word with 7 characters in a tie" do
+      words = ["pizzazz", "zzzzk", "pimp"]
+      Scrabble::Scoring.highest_score_from(words).must_equal "pizzazz"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:38]: Again trying the words in different order would be good.

File: specs/tilebag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    end
+  end
+
+  describe "#tiles_remaining" do
+    it "returns the number of tiles remaining in the bag" do
+      @tile.draw_tiles(4)
+      expected_length = 98 - 4
+
+      @tile.tiles_remaining.must_equal expected_length
+    end
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:63]: What happens when you draw all the tiles?

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
+require_relative 'spec_helper'
+require_relative '../lib/scoring'
+
+describe Scrabble::Scoring do
+  describe "#score" do
+    it "Calculates the score of one word" do
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:6]: What about Strings where not every character is a letter?

File: specs/board_spec.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+  before do
+    @b = Scrabble::Board.new
+    @right_turn = {word: 'TEST', column: 6, row: 6, direction: 'right'}
+    @down_turn = {word: 'TOO', column: 6, row: 6, direction: 'down'}
+  end
+
+  describe "#initialize" do
+    it "Creates a 15 x 15 blank board" do
+      @b.board.flatten.length.must_equal 225
```
Comment (CheezItMan) [inline specs/board_spec.rb:14]: What if it was of different dimensions what if it was 5 x 45?

Comment (CheezItMan) [reply specs/board_spec.rb:14]: Also what if the user tries to place a word outside the board?

**Review by CheezItMan — COMMENTED** [review]
Overall nicely done, it's neat to have a playable game.  I did leave some notes in your code here.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|  Good 	|
|   Both Teammates contributed to the codebase	|   Both teammates contributed commits, although it looks like Alix worked on the extras.  It also seems like Rana mostly wrote tests for code that Alix wrote.  Next time it would be a good idea to have both teammates contribute to the production code as well.  	|
|   Regular Commits with meaningful commit messages 	|   Good commit messages and fairly regular commits, nice!	|
|  **Scoring class**	|   Check	|
|  `score` method	|   Check	|
|   Uses appropriate data structure to store the letter score	|   Good use of a hash!	|
|   Appropriately handles edge cases	|   Basic edge cases tested, but you should also have tested things with the same array of words, but in different order to verify that the order of the words isn't affecting the results.  	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Check, but my comment above stands here too	|
|  `highest_scoring_word` method	|   Check and good use of enumerables	|
|   Appropriately handles edge cases	|   Check, but again you need some tests based on order.	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Check	|
|  **Player class**	|   Check	|
|  Uses an array to keep track of words played	|   Check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check	|
|  Returns `true` or `false` from `won?` method |  check  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   Some, but you need more testing on invalid words like, "some1"  |
|  **TileBag class**	|   Check	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Good use of an Array	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  Check, using shuffle 	|
| `tiles_remaining` method returns a count	|   Check	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   No testing here for when you draw all the tiles or attempt to draw more tiles than you should.	|
|  **Extras**	|  	|
|  Game Class	|  Well done!	|
|  Scrabble Game Running	|  Very neat!	|
|  Interactive Scrabble Board	| Nicely done	|

## Summary

Nicely done with lots of extras.  I do think you need to work a bit more on sharing coding responsibilities a bit more as both should be writing production code, and working a bit more collaboratively.  I also think you need to look a bit harder at testing throughly and looking at how your input can vary especially when the order of the input can affect the outcome, like in the Scoring class.  

I've left a few comments in your PR for specific feedback.
